### PR TITLE
Assignment2

### PR DESCRIPTION
**Pull request by grantdwoods**

### CODE REVIEW

**Comment by grantdwoods** [comment]
No unit tests for GiftImportService are ready at this time.

File: SecretSanta/SecretSanta.Import.Tests/Services/GiftImportServiceTest.cs [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+        [TestInitialize]
+        public void SetUpDirectory()
+        {
+            string tempDirPath = Path.GetTempPath();
+            dirPath = Path.Combine(tempDirPath + "SecretSantaGiftImport");
+
+            if (Directory.Exists(dirPath))
+                CleanUpDirectory();
+
+            Directory.CreateDirectory(dirPath);
```
Comment (KennethWhite) [inline SecretSanta/SecretSanta.Import.Tests/Services/GiftImportServiceTest.cs]: I like that you created a sub directory for your test data, I should have done this myself.

File: SecretSanta/SecretSanta.Import.Tests/Services/NameParsingServiceTest.cs [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+            string[] names = giftImportService.ParseHeader(": Woods, Grant");
+        }
+
+        [TestMethod]
+        [DataRow("Name:")]
+        [DataRow("Name:           ")]
+        [DataRow("Name: Grant")]
+        [DataRow("Name: Woods,")]
+        [DataRow("Name:  , Grant")]
+        [ExpectedException(typeof(ArgumentException), "first or last name missing.")]
```
Comment (KennethWhite) [inline SecretSanta/SecretSanta.Import.Tests/Services/NameParsingServiceTest.cs]: Nice use of data rows here to test the same exception.

File: SecretSanta/SecretSanta.Import/Services/NameParsingService.cs [hunk 3]
```diff
@@ -0,0 +1,95 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Text;
+
+namespace SecretSanta.Import.Services
+{
+    public class NameParsingService
```
Comment (KennethWhite) [inline SecretSanta/SecretSanta.Import/Services/NameParsingService.cs]: Your NameParsingSevice is really modular, and you were very thorough with addressing edge cases

File: SecretSanta/SecretSanta.Import.Tests/Services/GiftImportServiceTest.cs (hunk 1 above)
Comment (grantdwoods) [reply SecretSanta/SecretSanta.Import.Tests/Services/GiftImportServiceTest.cs]: I thought it was a good idea too! Though I should probably change how I deal with it based on what Mark said in class. Currently if that directory exists I just wipe it out. Mark said "What if someone else had a directory with that name there already". Very unlikely, but I understand his point!

File: SecretSanta/SecretSanta.Import/Services/NameParsingService.cs [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+        
+            string fullName = ParseFullName(header);
+            string[] names = SeparateFirstAndLast(fullName);
+
+            ValidateNamesArray(names);
+
+            return names;
+        }
+
+        private void ValidateHeaderStart(string header)
```
Comment (tdang2018) [inline SecretSanta/SecretSanta.Import/Services/NameParsingService.cs]: I really like the way you check whether the header is valid or not by using that method. It is super clearly to understand.

File: SecretSanta/SecretSanta.Import/Services/NameParsingService.cs [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+            {
+                throw new ArgumentException(nameof(header),
+                    "Header must start with \"Name: \"");
+            }
+        }
+        private void ValidateNamesArray(string[] names)
+        {
+            bool isValid = true;
+
+            if(names.Length !=2 )
```
Comment (tdang2018) [inline SecretSanta/SecretSanta.Import/Services/NameParsingService.cs]: I think here you can throw Exception if the length of names are not equal to 2 instead of set isValid to false. Then after that, you check if iValid is false, then throw Exception.

File: SecretSanta/SecretSanta.Import/Services/NameParsingService.cs [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+            return fullName.Trim();
+        }
+        private string[] SeparateFirstAndLast(string fullName)
+        {
+            string[] names;
+
+            if (fullName.Contains(','))
+            {
+                names = fullName.Split(',');
+                SwapToFirstLastFormat(names);
```
Comment (tdang2018) [inline SecretSanta/SecretSanta.Import/Services/NameParsingService.cs]: you call method SwapToFirstLastFormat() because I think we have 2 format Types:
Name: <FirstName> <LastName>
Name: <LastName>, <FirstName>
Is that correct?

File: SecretSanta/src/SecretSanta.Domain/Models/ApplicationDbContext.cs [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+        {
+
+        }
+
+        protected override void OnModelCreating(ModelBuilder modelBuilder)
+        {
+            modelBuilder.Entity<UserGroup>()
+                .HasKey(userGroup => new { userGroup.UserId, userGroup.GroupId });
+
+            modelBuilder.Entity<UserGroup>()
```
Comment (tdang2018) [inline SecretSanta/src/SecretSanta.Domain/Models/ApplicationDbContext.cs]: I do not understand the many to many for lamda expression. can you please explain about that? Thanks Grant

File: SecretSanta/SecretSanta.Import.Tests/Services/GiftImportServiceTest.cs [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+    public class GiftImportServiceTest
+    {
+        string dirPath;
+        string tmpFilePath;
+
+        [TestInitialize]
+        public void SetUpDirectory()
+        {
+            string tempDirPath = Path.GetTempPath();
+            dirPath = Path.Combine(tempDirPath + "SecretSantaGiftImport");
```
Comment (tdang2018) [inline SecretSanta/SecretSanta.Import.Tests/Services/GiftImportServiceTest.cs]: What SecretSantaGiftImport in this case mean? Is a file or Directory?

File: SecretSanta/SecretSanta.Import.Tests/Services/GiftImportServiceTest.cs [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+            ICollection<Gift> gifts = giftImportService.ImportGifts(tmpFilePath);
+            List<Gift> wishList = new List<Gift>(gifts);
+            Assert.AreEqual<string>("Grant", wishList[0].User.FirstName);
+            Assert.AreEqual<string>("Woods", wishList[1].User.LastName);
+            Assert.AreEqual<string>("Gift3", wishList[2].Title);
+
+            File.Delete(tmpFilePath);
+        }
+        [TestMethod]
+        public void ImportGifts_InvalidFile_ArgumentException()
```
Comment (tdang2018) [inline SecretSanta/SecretSanta.Import.Tests/Services/GiftImportServiceTest.cs]: This method is empty?

File: SecretSanta/SecretSanta.Import/Services/NameParsingService.cs (hunk 6 above)
Comment (grantdwoods) [reply SecretSanta/SecretSanta.Import/Services/NameParsingService.cs]: Exactly! I want to make sure that my array always contains [firstname, lastname] so I found the swap necessary.

File: SecretSanta/SecretSanta.Import.Tests/Services/GiftImportServiceTest.cs (hunk 8 above)
Comment (grantdwoods) [reply SecretSanta/SecretSanta.Import.Tests/Services/GiftImportServiceTest.cs]: I use dirPath to create the directory in the temp folder.

File: SecretSanta/SecretSanta.Import.Tests/Services/GiftImportServiceTest.cs (hunk 9 above)
Comment (grantdwoods) [reply SecretSanta/SecretSanta.Import.Tests/Services/GiftImportServiceTest.cs]: This was just a work in progress, pretty just a place holder at the ti.

File: SecretSanta/SecretSanta.Import/Services/NameParsingService.cs (hunk 5 above)
Comment (grantdwoods) [reply SecretSanta/SecretSanta.Import/Services/NameParsingService.cs]: Yeah I was thinking about trying not to repeat the same code too many times, but I agree the bool is probably redundant.

File: SecretSanta/src/SecretSanta.Domain/Models/ApplicationDbContext.cs (hunk 7 above)
Comment (grantdwoods) [reply SecretSanta/src/SecretSanta.Domain/Models/ApplicationDbContext.cs]: I do not remember the article I used, but Stokes and a fellow student both posted links on gitter for this. This one might help you out: https://blog.oneunicorn.com/2017/09/25/many-to-many-relationships-in-ef-core-2-0-part-1-the-basics/

File: SecretSanta/SecretSanta.Import/Services/NameParsingService.cs [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+            }
+            else
+            {
+                names = fullName.Split(" ");
```
Comment (MarkMichaelis) [inline SecretSanta/SecretSanta.Import/Services/NameParsingService.cs]: Your approach wouldn't work if I had "Name: Grant       Woods" but it would have been easy to allow and would be more robust against silly user errors.

File: SecretSanta/SecretSanta.Import.Tests/Services/NameParsingServiceTest.cs [hunk 11]
```diff
@@ -0,0 +1,101 @@
+using Microsoft.VisualStudio.TestTools.UnitTesting;
+using SecretSanta.Import.Services;
+using System;
+
+namespace SecretSanta.Import.Tests
+{
+    [TestClass]
+    public class NameParsingServiceTest
+    {
+        public NameParsingService CreateDefaultImporter()
```
Comment (MarkMichaelis) [inline SecretSanta/SecretSanta.Import.Tests/Services/NameParsingServiceTest.cs]: I'm not sure you gain much by having a method that only calls a constructor?  It forces the reader to look up what the function does rather than just inline it and see it.  Obviously, if there was or was likely to be more complexity, perhaps.
An alternative approach would be to have a NamingParsingService property and to set it in TestInitialize.

File: SecretSanta/SecretSanta.Import.Tests/Services/GiftImportServiceTest.cs [hunk 12]
```diff
@@ -0,0 +1,126 @@
...
+            DirectoryInfo dirInfo = new DirectoryInfo(dirPath);
+            foreach (FileInfo file in dirInfo.GetFiles())
+            {
+                File.Delete(file.FullName);
+            }
+            Directory.Delete(dirPath);
+        }
+
+        [TestMethod]
+        [DataRow(new[] { "Name: Grant Woods", "Gift1", "Gift2", "Gift3" })]
```
Comment (MarkMichaelis) [inline SecretSanta/SecretSanta.Import.Tests/Services/GiftImportServiceTest.cs]: I like the use of the DataRow attribute.

File: SecretSanta/SecretSanta.Import.Tests/Services/GiftImportServiceTest.cs [hunk 13]
```diff
@@ -0,0 +1,126 @@
...
+        {
+            GiftImportService giftImportService = new GiftImportService();
+
+            File.WriteAllLines(tmpFilePath, fileInputArray);
+            try
+            {
+                ICollection<Gift> gifts = giftImportService.ImportGifts(tmpFilePath);
+                Assert.Fail();
+            }
+            catch (Exception exception)
```
Comment (MarkMichaelis) [inline SecretSanta/SecretSanta.Import.Tests/Services/GiftImportServiceTest.cs]: Again... no, just catch an ArgumentException.  
Catching `Exception` and not re-throwing is almost always wrong.  It suggests you could appropriately handle any exception which is unlikely in general.

File: SecretSanta/SecretSanta.Import.Tests/Services/GiftImportServiceTest.cs [hunk 14]
```diff
@@ -0,0 +1,126 @@
...
+            Assert.AreEqual<string>("Woods", wishList[1].User.LastName);
+            Assert.AreEqual<string>("Gift3", wishList[2].Title);
+
+            File.Delete(tmpFilePath);
+        }
+
+        [TestMethod]
+        [DataRow(new[] { "Name: Grant Woods", "", "", "" })]
+        [DataRow(new[] { "Name: Grant Woods", "            ", " ", "" })]
+        public void ImportGifts_ValidHeaderMissingGifts_ThrowsArgumentException(string[] fileInputArray)
```
Comment (MarkMichaelis) [inline SecretSanta/SecretSanta.Import.Tests/Services/GiftImportServiceTest.cs]: The code os convoluted.  Using an ExpectedException attribute would help significantly.

**Comment by MarkMichaelis** [comment]
Grant.... I really like your implementation.  Nice job!!
